components/scream/src/share/util/scream_column_ops.hpp

```diff
@@ -44,26 +45,42 @@ namespace scream {
  *    auto prod = [&](const int k)->pack_type { return x(k)*y(k); }
  *    col_ops::compute_midpoint_values(team,nlevs,prod,output);
  *
- *  Note: all methods have a different impl, depending on whether PackSize>1.
- *        The if statement is evaluated at compile-time, so there is no run-time
- *        penalization. The only requirement is that both branches must compile.
+ *  Notes:
+ *   - all methods have a different impl, depending on whether PackSize>1.
+ *     The if statement is evaluated at compile-time, so there is no run-time
+ *     penalization. The only requirement is that both branches must compile.
+ *   - some methods accept a non-type template argument of type CombineMode.
+ *     This argument can be used to specify how the result of the calculation
+ *     should be written in the output view. E.g., if CM=ScaleUpdate, the output
+ *     view y will be updated as y = beta*y + alpha*f(x). The values alpha
+ *     and beta are used only if CM needs them, and an error is thrown if the
+ *     user specifies non-trivial alpha/beta when they are not needed.
+ *     See scream_combine_ops.hpp for more details.
  *
  *  RECALL: k=0 is the model top, while k=num_mid_levels+1 is the surface!
  */
 
-template<typename DeviceType, typename ScalarType, int PackSize>
+template<typename DeviceType, typename ScalarType>
 class ColumnOps {
 public:
   // Expose template params
   using device_type = DeviceType;
   using scalar_type = ScalarType;
-  using pack_type   = ekat::Pack<scalar_type,PackSize>;
 
-  // Pack info
-  enum : int {
-    pack_size = PackSize,
-  };
-  using pack_info = ekat::PackInfo<pack_size>;
+  template<int PackSize>
+  using pack_type = ekat::Pack<scalar_type,PackSize>;
+
+  template<typename ScalarT>
+  static constexpr bool is_simd () {
+    return ekat::ScalarTraits<ScalarT>::is_simd;
+  }
+
+  template<typename ScalarT>
+  KOKKOS_FUNCTION
+  static constexpr int pack_size () {
```

Comment:
Is this for computing the pack size of another type of Pack for easier conversion between packs of different types?

Reply:
Well, `ScalarT` may or may not be a Pack. You can call any of these methods with ScalarT=Real or ScalarT=Pack<Real,N>. So I needed a helper function that gave me "the number of Real's" in a ScalarT.

I think eventually I should change this, doing what I did in PhysicsFunctions, namely, do not template the class on the scalar type, but template only the individual functions. I think that might make the code a tad cleaner, but it might take some coding time, so I'm punting on this.

Reply:
Good to know that `ScalarT` describes both actual scalars and packs. I think of packs as a "vector" type, so this wasn't obvious to me.

Reply:
Yeah, tbh, I am not fond of how we use the term "scalar" in scream/ekat. I am thinking that `ValueType` is often more appropriate than `ScalarType`.